snap/squashfs/stat.go

```diff
@@ -0,0 +1,330 @@
+	}
+
+	if err := st.parseTime(raw[p : p+16]); err != nil {
+		return nil, err
+	}
+
+	p += 16
+
+	if raw[p] != ' ' {
+		return nil, errBadLine(raw)
+	}
+	p++
+	if raw[p] != '.' {
+		return nil, errBadLine(raw)
+	}
+	p++
+	if len(raw[p:]) == 0 {
+		st.path = "/"
+	} else {
+		st.path = string(raw[p:])
+	}
+
+	return st, nil
+}
+
+type statError struct {
+	part string
+	raw  []byte
+}
+
+func (e statError) Error() string {
+	return fmt.Sprintf("cannot parse %s: %q", e.part, e.raw)
+}
+
+func errBadLine(raw []byte) statError {
+	return statError{
+		part: "line",
+		raw:  raw,
+	}
+}
+
+func errBadMode(raw []byte) statError {
+	return statError{
+		part: "mode",
+		raw:  raw,
+	}
+}
+
+func errBadOwner(raw []byte) statError {
+	return statError{
+		part: "owner",
+		raw:  raw,
+	}
+}
+
+func errBadNode(raw []byte) statError {
+	return statError{
+		part: "node",
+		raw:  raw,
+	}
+}
+
+func errBadSize(raw []byte) statError {
+	return statError{
+		part: "size",
+		raw:  raw,
+	}
+}
+
+func (st *stat) parseTime(raw []byte) error {
+	t, err := time.ParseInLocation("2006-01-02 15:04", string(raw), time.Local)
+	if err != nil {
+		return err
+	}
+
+	st.mtime = t
+
+	return nil
+}
+
+func (st *stat) parseMode(raw []byte) error {
+	switch raw[0] {
+	case '-':
+		// 0
+	case 'd':
+		st.mode |= os.ModeDir
+	case 's':
+		st.mode |= os.ModeSocket
+	case 'c':
+		st.mode |= os.ModeCharDevice
+	case 'b':
+		st.mode |= os.ModeDevice
+	case 'p':
+		st.mode |= os.ModeNamedPipe
+	case 'l':
+		st.mode |= os.ModeSymlink
+	default:
+		return errBadMode(raw)
+	}
+
+	for i := 0; i < 3; i++ {
+		m, me := modeFromTriplet(raw[1+3*i:4+3*i], uint(2-i))
+		if me != nil {
+			return me
+		}
+		st.mode |= m
+	}
+
+	return nil
+}
+
+func (st *stat) parseOwner(raw []byte) (int, error) {
+	var p, ui, uj, gi, gj int
+
+	// first check it's sane (starts with space, and then at least two non-space chars
+	if raw[0] != ' ' || raw[1] == ' ' || raw[2] == ' ' {
+		return 0, errBadLine(raw)
+	}
+
+	ui = 1
+out:
+	for p = ui; p < 40; p++ {
+		switch raw[p] {
+		case '/':
+			uj = p
+			gi = p + 1
+		case ' ':
+			gj = p
+			break out
+		}
+	}
+
+	if uj == gj || gi >= p {
+		return 0, errBadOwner(raw)
+	}
+	st.user, st.group = string(raw[ui:uj]), string(raw[gi:gj])
+
+	return p, nil
+}
+
+func modeFromTriplet(trip []byte, shift uint) (os.FileMode, error) {
+	var mode os.FileMode
+	high := false
+	if len(trip) != 3 {
+		panic("bad triplet length")
+	}
+	switch trip[0] {
+	case '-':
+		// 0
+	case 'r':
+		mode |= 4
+	default:
+		return 0, errBadMode(trip)
+	}
+	switch trip[1] {
+	case '-':
+		// 0
+	case 'w':
+		mode |= 2
+	default:
+		return 0, errBadMode(trip)
+	}
+	switch trip[2] {
+	case '-':
+		// 0
+	case 'x':
+		mode |= 1
+	case 'S', 'T':
+		high = true
+	case 's', 't':
+		mode |= 1
+		high = true
+	default:
+		return 0, errBadMode(trip)
+	}
+
+	mode <<= 3 * shift
+	if high {
+		mode |= (01000 << shift)
+	}
+	return mode, nil
+}
+
+func (st *stat) parseSize(raw []byte) (int, error) {
```

Comment:
Maybe a comment what the expected inputs are (once with  isNode, one without) to make it easier to follow the common case?

Reply:
And is `isNode` tested?